lsp/nls/src/requests/completion.rs

```diff
@@ -21,56 +21,79 @@ use crate::{
     trace::{Enrich, Trace},
 };
 
-/// Find record fields for an item with the specified id.
-fn find_record_fields(linearization: &Completed, id: usize) -> Option<Vec<Ident>> {
+// General ideas:
+// A path is the reverse of the list of identifiers that make up a record indexing operation.
+// e.g if we have a.b.c, the associated path would be vec![c, b, a]
+// Paths are used to guide the completion engine to handle nested records.
+// They also generalize the single record indexing case.
+
+/// Find the record field associated with a particular ID in the linearization
+/// using lexical scoping rules.
+fn find_fields_from_term_kind(
+    linearization: &Completed,
+    id: usize,
+    path: &mut Vec<Ident>,
+) -> Option<Vec<Ident>> {
     let item = linearization.get_item(id)?;
     match item.kind {
-        TermKind::Record(ref fields) if item.id == id => Some(fields.keys().cloned().collect()),
-        TermKind::Declaration(_, _, ValueState::Known(new_id))
-        | TermKind::Usage(UsageState::Resolved(new_id))
-            if item.id == id =>
-        {
-            find_record_fields(linearization, new_id)
+        TermKind::Record(ref fields) => {
+            if path.is_empty() {
+                Some(fields.keys().cloned().collect())
+            } else {
+                let name = path.pop()?;
+                let new_id = fields.get(&name)?;
+                find_fields_from_term_kind(&linearization, *new_id, path)
+            }
+        }
+        TermKind::RecordField {
+            value: ValueState::Known(new_id),
+            ..
+        }
+        | TermKind::Declaration(_, _, ValueState::Known(new_id))
+        | TermKind::Usage(UsageState::Resolved(new_id)) => {
+            find_fields_from_term_kind(linearization, new_id, path)
         }
         _ => None,
     }
 }
 
-/// Find a record contract of the item with the specified id.
-fn find_contract_record_fields(linearization: &Completed, id: usize) -> Option<Vec<Ident>> {
+/// Find the record fields associated with an ID in the linearization using
+/// its contract information.
+fn find_fields_from_contract(
+    linearization: &Completed,
+    id: usize,
+    path: &mut Vec<Ident>,
+) -> Option<Vec<Ident>> {
     let item = linearization.get_item(id)?;
     match &item.meta {
-        Some(MetaValue { contracts, .. }) if item.id == id => {
+        Some(MetaValue { contracts, .. }) => {
             contracts.iter().find_map(|contract| match &contract.types {
-                Types(AbsType::Record(row)) => Some(extract_ident(&row)),
-                Types(AbsType::Flat(term)) => {
```

Comment:
@ebresafegaga why do we get rid of the code that handles the case where the contract is a record contract (and not a record type)? It looks like we're removing a feature here.

Also, there may be a type annotation inside `meta.types`. In general, if you want **all** the annotations on a value, mixing contracts and the type annotation, you should do something like `meta.types.iter().chain(meta.contracts.iter())`

Reply:
This was on purpose. I removed it completely, to be properly implemented with item `3` in #877. 

It should be able to handle a mixture of types and contracts: 
```nickel
let term | {a | {b : { c | { d : { f : Str } } }} } = { ... } in a. 
```
So, the implementation for item `3` will do this